upgrade/v15xto160/upgrade.go

```diff
@@ -160,6 +199,127 @@ func upgradeVolumeAttachments(namespace string, lhClient *lhclientset.Clientset,
 	return nil
 }
 
+func upgradeBackupTargets(namespace string, lhClient *lhclientset.Clientset, resourceMaps map[string]interface{}) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"upgrade backup target failed")
+	}()
+
+	backupTargetMap, err := upgradeutil.ListAndUpdateBackupTargetsInProvidedCache(namespace, lhClient, resourceMaps)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list all existing Longhorn BackupTargets during the Longhorn BackupTarget upgrade")
+	}
+
+	backupTarget, exist := backupTargetMap[oldDefaultBackupTargetName]
+	if !exist {
+		return errors.Wrapf(err, "failed to get old default backup target during the Longhorn BackupTarget upgrade")
+	}
+
+	backupTarget.Spec.Default = true
+
+	return nil
+}
+
+func upgradeBackupVolumes(namespace string, lhClient *lhclientset.Clientset, resourceMaps map[string]interface{}) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"upgrade backup volume failed")
+	}()
+
+	backupTarget, err := getOldDefaultBackupTarget(namespace, lhClient, resourceMaps)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get old default backup target during the Longhorn BackupVolume upgrade")
+	}
+
+	backupVolumeMap, err := upgradeutil.ListAndUpdateBackupVolumesInProvidedCache(namespace, lhClient, resourceMaps)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to list all existing Longhorn BackupVolumes during the Longhorn BackupVolume upgrade")
+	}
+
+	for _, bv := range backupVolumeMap {
+		if bv.Spec.BackupTargetName == "" {
+			bv.Spec.BackupTargetName = backupTarget.Name
+		}
+		if bv.Spec.BackupTargetURL == "" {
+			bv.Spec.BackupTargetURL = backupTarget.Spec.BackupTargetURL
+		}
+		if bv.Spec.VolumeName == "" {
+			bv.Spec.VolumeName = bv.Name
+		}
+		if bv.Labels == nil {
+			bv.Labels = make(map[string]string)
+		}
+		_, exists := bv.Labels[types.LonghornLabelBackupVolumeCRName]
+		if !exists {
+			bv.Labels[types.LonghornLabelBackupVolumeCRName] = bv.Spec.VolumeName + "-" + bv.Spec.BackupTargetName
+		}
+	}
+
+	return nil
+}
+
+func upgradeBackups(namespace string, lhClient *lhclientset.Clientset, resourceMaps map[string]interface{}) (err error) {
```

Comment:
Just to make sure
1. for those old BackupVolume we will keep the name 
2. for newly create BackupTarget, the BackupVolumes of it will still have the name "VolumeName-BackupTargetName" so it won't have conflict
3. we now also store the BackupTarget info in BackupVolume's label, and we use that to list BackupVolumes with specific BackupTarget

Is that correct?

Reply:
1. Yes
2. Yes
3. By purpose, Yes to what you mentioned.
Usually, we get a backup volume CR by matching the composite name `VolumeName-BackupTargetName` to the label `LonghornLabelBackupVolumeCRName (backup-volume-cr-name)` of the backup volume CR. We can find a backup volume by a volume name and a backup target name for sure.